core/iwasm/fast-jit/fe/jit_emit_numberic.c

```diff
@@ -472,16 +472,104 @@ jit_compile_op_i64_bitwise(JitCompContext *cc, IntBitwise bitwise_op)
     return false;
 }
 
+static bool
+compile_int_shift(JitCompContext *cc, IntShift shift_op, bool is_i32)
+{
+#if defined(BUILD_TARGET_X86_64) || defined(BUILD_TARGET_AMD_64)
+    JitReg ecx_hreg = jit_codegen_get_hreg_by_name("ecx");
+    JitReg rcx_hreg = jit_codegen_get_hreg_by_name("rcx");
+#endif
+    JitReg left, right, mod_right, res;
+
+    POP_INT(right);
+    POP_INT(left);
+
+    if (IS_CONST_ZERO(right)) {
+        res = left;
+        goto shortcut;
+    }
+
+    /* right modulo N */
+    if (is_i32) {
+        mod_right = jit_cc_new_reg_I32(cc);
+        GEN_INSN(REM_U, mod_right, right, NEW_CONST(I32, 32));
+    }
+    else {
+        mod_right = jit_cc_new_reg_I64(cc);
+        GEN_INSN(REM_U, mod_right, right, NEW_CONST(I64, 64));
+    }
+
+    if (IS_CONST_ZERO(right)) {
+        res = left;
```

Comment:
Code is duplicated with L487-489, do you want to check `IS_CONST_ZERO(mod_right)`?
But it is not const reg. How about change L487-489 to:
```
if (jit_reg_is_const_val(right)) {
    if ((is_i32 && jit_cc_get_const_I32(cc, right) % 32 == 0))
        || (!is_i32 && jit_cc_get_const_I64(cc, right) % 64 == 0)) {
        res = left;
        goto shortcut;
      }
  }
```

Reply:
Yes. I should check `mod_right` in the second `IS_CONST_ZERO`. But I am a little confused about that the real thing we should do is check the value of `JitReg`, no matter it is a const or not. I mean maybe we should create some `conditional branches` to handle that?

Reply:
My understanding is there is no need to check the value of `JitReg right`, no matter what the `right` value is, we didn't check it in interpreter/AOT. The spec defines:
```
ishl𝑁(𝑖1, 𝑖2)
• Let 𝑘 be 𝑖2 modulo 𝑁.
• Return the result of shifting 𝑖1 left by 𝑘 bits, modulo 2𝑁.
ishl𝑁(𝑖1, 𝑖2) = ibits− 𝑁1(𝑑𝑁 2 −𝑘 0𝑘) (if ibits𝑁(𝑖1) = 𝑑𝑘 1 𝑑𝑁 2 −𝑘 ∧ 𝑘 = 𝑖2 mod 𝑁)
```
It doesn't mention that we should add any check or to throw exception.